### PR TITLE
Remedies for Scorecard Vulnerabilities [Wave 1]

### PR DESCRIPTION
**Pull request by daffafaizan**
### Overview
This pull request introduces the first wave remedies for the vulnerabilities and warnings from OSSF Scorecard. Other than that, we added a license and a security guideline. This pull request also includes changes and hot-fixes to the `main` branch; also a back-merge pull request from `main` to `dev`.

### Vulnerabilities Remedied
1. Pinned-dependencies for:
   a. build.yml
   b. dev-deployment.yml
   c. prod-deployment.yml
   d. Dockerfile
   
   by changing (for example):
   ```
   uses: actions/checkout@v4
   ```
   to this:
   ```
   uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
   ```
   
2. Restrict permissions for:
   a. dev-deployment.yml
   b. prod-deployment.yml
   
   by adding:
      ```
      permissions:
         contents: read
      ```
      
3. Removed information exposure in exception handling
   by changing (for example):
   ```
    except ClientError as e:
        return JSONResponse(
            status_code=http.HTTPStatus.BAD_REQUEST,
            content=f"Error: Failed to delete transcription job: {e}",
        )
   ```
   to a more generic error response, like this:
   ```
    except ClientError as e:
        return JSONResponse(
            status_code=http.HTTPStatus.BAD_REQUEST,
            content=f"Error: Failed to delete transcription job",
        )
   ```
   
### Additions for Security
1. Added SECURITY.md containing security guidelines.
2. Added TUTORIAL.md regarding how to clone and run the project.
3. Modified README.md containing a general description about the project and it's features.

### Unrelated Additional Changes
1. Changed method of access to headers for renewing access_token
2. Added endpoint to check if email exists or not at `v1/auth/email-exist` with the request body
```
{
   "email" : "your@email.com"
}
```
3. Added `title` field to generate flashcards request schema so it can be saved to the model
4. Added `set_id` field to generate flashcards response schema so it can be used in the front-end

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 
 # Select base image - Python 3.12
-FROM python:3.12-slim
+FROM python:3.12-slim@sha256:afc139a0a640942491ec481ad8dda10f2c5b753f5c969393b12480155fe15a63
```
Comment (jeremyalv) [inline Dockerfile:3]: Let's keep the generic 3.12 python version, can just revert this change.

File: .github/workflows/prod-deployment.yml [hunk 2]
```diff
@@ -21,11 +24,11 @@ jobs:
     
     steps:
       - name: Checkout Sources
-        uses: actions/checkout@v4
+        uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
```
Comment (jeremyalv) [inline .github/workflows/prod-deployment.yml:27]: For `dev-deployment` and `prod-deployment`, let's not change anything to avoid unnecessary CI/CD debugging. (If it works, don't change it unless we really have to).

File: README.md
Comment (jeremyalv) [inline README.md:1]: Nice update, thanks for this

File: Dockerfile (hunk 1 above)
Comment (daffafaizan) [reply Dockerfile:3]: Alright, I've changed it.

File: .github/workflows/prod-deployment.yml (hunk 2 above)
Comment (daffafaizan) [reply .github/workflows/prod-deployment.yml:27]: Didn't think of that. I'll change it back.

**Review by jeremyalv — APPROVED** [review]

**Review by jeremyalv — APPROVED** [review]
